### PR TITLE
TF2 examples

### PR DESCRIPTION
**Pull request by vandanavk**
### Description of changes:
Adding one example script each for TF 2.x keras fit() eager , TF 2.x keras fit() non-eager, TF 2.x GradientTape.

The following table describes what tensors the examples save

|                                     | keras fit() eager | keras fit() non-eager | GradientTape |
|-------------------------------------|-------------------|-----------------------|--------------|
| loss                                | yes               | yes                   | yes          |
| metrics                             | yes               | yes                   | yes          |
| weights                             | yes               | yes                   | yes          |
| biases                              | yes               | yes                   | yes          |
| gradients                           | no                | yes                   | yes          |
| input/output to model               | no                | no                    | no           |
| input/output of intermediate layers | no                | yes                   | no           |
| optimizer variables                 | no                | yes                   | no           |
| arbitrary scalar                                    | yes                  | yes                      |  yes            |
|                                     |                   |                       |              |

#### Style and formatting:

I have run `pre-commit install` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Vikas-kum** [comment]
1. Example to save arbitrary scalar . (if you can please Make sure that you can see it in minerva & studio)
input/output to model - did you try it for non-eager mode ?

File: examples/tensorflow2/scripts/tf_keras_gradienttape.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
+"""
+This script is a ResNet training script which uses Tensorflow's Keras interface.
+It has been orchestrated with SageMaker Debugger hook to allow saving tensors during training.
+Here, the hook has been created using its constructor to allow running this locally for your experimentation.
+When you want to run this script in SageMaker, it is recommended to create the hook from json file.
+Please see scripts in either 'sagemaker_byoc' or 'sagemaker_official_container' folder based on your use case.
```
Comment (Vikas-kum) [inline examples/tensorflow2/scripts/tf_keras_gradienttape.py:6]: relative path to folder.

File: examples/tensorflow2/scripts/tf_keras_gradienttape.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import numpy as np
+import tensorflow.compat.v2 as tf
+from tensorflow.keras.applications.resnet50 import ResNet50
+from tensorflow.keras.datasets import cifar10
+from tensorflow.keras.utils import to_categorical
+
+# First Party
+import smdebug.tensorflow as smd
+
+
```
Comment (Vikas-kum) [inline examples/tensorflow2/scripts/tf_keras_gradienttape.py:22]: QQ: is gradient tape eager mode?

File: examples/tensorflow2/scripts/tf_keras_gradienttape.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    num_training_samples = x_train.shape[0]
+
+    for epoch in range(n_epochs):
+        print("Epoch %d/%d" % (epoch + 1, n_epochs))
+        progBar = tf.keras.utils.Progbar(num_training_samples, stateful_metrics="accuracy")
+        for idx, (data, labels) in enumerate(dataset):
+            dataset_labels = labels
+            labels = to_categorical(labels, 10)
+            # wrap the tape so the hook can identify the gradients, parameters, loss
+            with hook.wrap_tape(tf.GradientTape()) as tape:
```
Comment (Vikas-kum) [inline examples/tensorflow2/scripts/tf_keras_gradienttape.py:51]: more comments.

File: examples/tensorflow2/README.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# Examples
+## Example notebooks
+Please refer to the example notebooks in [Amazon SageMaker Examples repository](https://github.com/awslabs/amazon-sagemaker-examples/tree/master/sagemaker-debugger)
```
[the remark this answers is not in the document]
Comment (vandanavk) [reply examples/tensorflow2/README.md:3]: https://github.com/awslabs/amazon-sagemaker-examples/pull/1102

File: smdebug/tensorflow/keras.py [hunk 5]
```diff
@@ -726,11 +726,18 @@ def run(*args, **kwargs):
...
-                        tensor_name="gradients/" + layer + "Grad",
-                        tensor_value=g,
-                        check_before_write=True,
-                    )
+                    # Adding a check to make sure gradients are not None.
+                    # gradients may be None if user tries to compute gradients for
+                    # non-training variable when using model.variables instead of
+                    # model.trainable_variables in tape.gradient().
+                    # model.variables includes trainable and non-trainable
+                    # variables.
```
[the remark this answers is not in the document]
Comment (vandanavk) [reply smdebug/tensorflow/keras.py:734]: This is not expected to hit as users would normally use `model.trainable_variables` in `tape.gradient()`. But if users use `model.variables` unintentionally, like I did, then they will face an issue.

File: examples/tensorflow2/scripts/tf_keras_fit_eager.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
+"""
+This script is a ResNet training script which uses Tensorflow's Keras interface.
+It has been orchestrated with SageMaker Debugger hook to allow saving tensors during training.
+Here, the hook has been created using its constructor to allow running this locally for your experimentation.
+When you want to run this script in SageMaker, it is recommended to create the hook from json file.
+Please see scripts in either 'sagemaker_byoc' or 'sagemaker_official_container' folder based on your use case.
```
[the remark this answers is not in the document]
Comment (vandanavk) [reply examples/tensorflow2/scripts/tf_keras_fit_eager.py:6]: Have to remove this comment. I structured the directory similar to mxnet and pytorch. This comment is coming from tensorflow directory

File: examples/tensorflow2/scripts/tf_keras_gradienttape.py (hunk 1 above)
Comment (vandanavk) [reply examples/tensorflow2/scripts/tf_keras_gradienttape.py:6]: Have to remove this comment. I structured the directory similar to mxnet and pytorch. This comment is coming from tensorflow directory

File: examples/tensorflow2/scripts/tf_keras_gradienttape.py (hunk 2 above)
Comment (vandanavk) [reply examples/tensorflow2/scripts/tf_keras_gradienttape.py:22]: yes

**Comment by vandanavk** [comment]
> 1. Example to save arbitrary scalar . (if you can please Make sure that you can see it in minerva & studio)
>    input/output to model - did you try it for non-eager mode ?

I printed all the tensors from the trial, but I don't see model input/output. checked TF 1.15 non-eager, TF 2.1 non-eager as well. do we support this?

File: examples/tensorflow2/scripts/tf_keras_fit_non_eager.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    hook = smd.KerasHook(
+        out_dir=opt.out_dir,
+        include_collections=[
+            "weights",
+            "biases",
+            "default",
+            "gradients",
+            "optimizer_variables",
+            "inputs",
+            "outputs",
```
Comment (Vikas-kum) [inline examples/tensorflow2/scripts/tf_keras_fit_non_eager.py:74]: is this pre-defined collection. I don't see this in https://github.com/awslabs/sagemaker-debugger/blob/master/docs/api.md#collection ? 
Can you check if we pre-define this, and docs need to be modified?

File: examples/tensorflow2/scripts/tf_keras_gradienttape.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        progBar = tf.keras.utils.Progbar(num_training_samples, stateful_metrics="accuracy")
+        for idx, (data, labels) in enumerate(dataset):
+            dataset_labels = labels
+            labels = to_categorical(labels, 10)
+            # wrap the tape so the hook can identify the gradients, parameters, loss
+            # wrapping the tape ensures that smdebug's wrapper around functions of
+            # the tape object - push_tape(), pop_tape(), gradient(), will setup the writers of
+            # the debugger and save tensors that are provided as input to gradient() (trainable
+            # variables and loss), output of gradient() (gradients).
+            with hook.wrap_tape(tf.GradientTape()) as tape:
```
Comment (Vikas-kum) [inline examples/tensorflow2/scripts/tf_keras_gradienttape.py:54]: Ideally we can do a ZCC for this? by modifying GradientTape function correct?

File: examples/tensorflow2/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (Vikas-kum) [reply examples/tensorflow2/README.md:3]: we should have specific link in the doc. It will be easier for user to find it rather than guessing. May be a follow up PR after sagemaker-example is merged. 
tensorflow2/tensorflow2_zero_code_change/tf2-keras-default-container.ipynb  , for now you can point out to this script.

File: examples/tensorflow2/scripts/tf_keras_fit_non_eager.py (hunk 7 above)
Comment (Vikas-kum) [reply examples/tensorflow2/scripts/tf_keras_fit_non_eager.py:74]: also comment with docs link about thiese default collections

File: examples/tensorflow2/scripts/tf_keras_gradienttape.py (hunk 8 above)
Comment (vandanavk) [reply examples/tensorflow2/scripts/tf_keras_gradienttape.py:54]: we cant do ZCC. ZCC doesn't support GradientTape yet.

File: examples/tensorflow2/scripts/tf_keras_fit_non_eager.py (hunk 7 above)
Comment (vandanavk) [reply examples/tensorflow2/scripts/tf_keras_fit_non_eager.py:74]: its not pre-defined. these are the collections that are saved when save_all=True

File: examples/tensorflow2/scripts/tf_keras_fit_eager.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    parser.add_argument("--save_interval", type=int, default=500)
+    opt = parser.parse_args()
+
+    model = ResNet50(weights=None, input_shape=(32, 32, 3), classes=10)
+
+    ##### Enabling SageMaker Debugger ###########
+    # creating hook
+    hook = smd.KerasHook(
+        out_dir=opt.out_dir,
+        include_collections=["weights", "biases", "default"],
```
[the remark this answers is not in the document]
Comment (vandanavk) [reply examples/tensorflow2/scripts/tf_keras_fit_eager.py:66]: done in the latest commit @Vikas-kum

**Comment by vandanavk** [comment]
> > 1. Example to save arbitrary scalar . (if you can please Make sure that you can see it in minerva & studio)
> >    input/output to model - did you try it for non-eager mode ?
> 
> will check save_scalar and update here.
> 
> for input/output to the model,
> I printed all the tensors from the trial, but I don't see model input/output. checked TF 1.15 non-eager, TF 2.1 non-eager as well. do we support this?

keras fit eager mode. scalar/epoch and scalar/batch_size have been saved using hook.save_scalar API

Screen Shot 2020-04-16 at 11 49 14 PM

File: examples/tensorflow2/scripts/tf_keras_fit_non_eager.py (hunk 7 above)
Comment (Vikas-kum) [reply examples/tensorflow2/scripts/tf_keras_fit_non_eager.py:82]: As discussed offline, there is something off about this - 
1) Either we have code or documentation problem ( input is not there in documented collection) 
2) Please check what gets saved for input & output  collection for all frameworks . We may need to tighten docs accordingly.

File: examples/tensorflow2/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (Vikas-kum) [reply examples/tensorflow2/README.md:3]: Can you create a issue to track this.

**Comment by vandanavk** [comment]
Removed inputs from the non-eager example as they are not being saved.

Screen Shot 2020-04-17 at 2 14 13 PM
